File: library/src/main/java/se/ingenuity/markdownview/util/ResolvedAttributes.java
package se.ingenuity.markdownview.util;

import static se.ingenuity.markdownview.R.styleable.MarkdownTextView;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdBlockQuoteStyle;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdCodeBlockBackgroundColor;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdCodeBlockMargin;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdCodeBlockTextColor;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdCodeStyle;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdEmphasisStyle;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdFencedCodeBlockStyle;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdH1PostStyle;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdH1PreStyle;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdH1Style;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdH2PostStyle;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdH2PreStyle;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdH2Style;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdH3PostStyle;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdH3PreStyle;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdH3Style;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdH4PostStyle;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdH4PreStyle;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdH4Style;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdH5PostStyle;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdH5PreStyle;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdH5Style;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdH6PostStyle;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdH6PreStyle;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdH6Style;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdIndentedCodeBlockStyle;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdLinkStyle;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdLinkUnderlined;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdListItemColor;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdOrderedListItemStyle;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdParagraphStyle;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdStrongEmphasisStyle;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdThematicBreakColor;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdThematicBreakHeight;
import static se.ingenuity.markdownview.R.styleable.MarkdownTextView_mdUnorderedListItemStyle;
import static se.ingenuity.markdownview.util.Constants.ID_NULL;
import static se.ingenuity.markdownview.util.Constants.UNDEFINED_DIMEN;

import android.content.Context;
import android.content.res.ColorStateList;
import android.content.res.TypedArray;
import android.util.AttributeSet;

import androidx.annotation.AttrRes;
import androidx.annotation.Dimension;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StyleRes;
import androidx.annotation.StyleableRes;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

class ResolvedAttributes {
    @NonNull
    final List<StyleGroup> headingStyleGroups;
    @NonNull
    final StyleGroup emphasisStyleGroup;
    @NonNull
    final StyleGroup strongEmphasisStyleGroup;
    @NonNull
    final StyleGroup blockQuoteStyleGroup;
    @NonNull
    final StyleGroup codeStyleGroup;
    @NonNull
    final StyleGroup fencedCodeBlockStyleGroup;
    @NonNull
    final StyleGroup indentedCodeBlockStyleGroup;
    @NonNull
    final StyleGroup orderedListItemStyleGroup;
    @NonNull
    final StyleGroup unorderedListItemStyleGroup;
    @NonNull
    final StyleGroup paragraphStyleGroup;
    @NonNull
    final StyleGroup linkStyleGroup;

    @Nullable
    final Boolean linkUnderlined;

    @Nullable
    final ColorStateList listItemColor;
    @Nullable
    final ColorStateList codeBlockBackgroundColor;
    @Nullable
    final ColorStateList codeBlockTextColor;
    @Dimension
    final int codeBlockMargin;
    @Nullable
    final ColorStateList thematicBreakColor;
    @Dimension
    final int thematicBreakHeight;

    ResolvedAttributes(@NonNull Context context,
                       @Nullable AttributeSet attrs,
                       @AttrRes int defStyleAttr,
                       @StyleRes int defStyleRes) {
        final TypedArray a = context.obtainStyledAttributes(
                attrs,
                MarkdownTextView,
                defStyleAttr,
                defStyleRes);

        headingStyleGroups = Collections.unmodifiableList(Arrays.asList(
                StyleGroup.of(a,
                        MarkdownTextView_mdH1PreStyle,
                        MarkdownTextView_mdH1Style,
                        MarkdownTextView_mdH1PostStyle),
                StyleGroup.of(a,
                        MarkdownTextView_mdH2PreStyle,
                        MarkdownTextView_mdH2Style,
                        MarkdownTextView_mdH2PostStyle),
                StyleGroup.of(a,
                        MarkdownTextView_mdH3PreStyle,
                        MarkdownTextView_mdH3Style,
                        MarkdownTextView_mdH3PostStyle),
                StyleGroup.of(a,
                        MarkdownTextView_mdH4PreStyle,
                        MarkdownTextView_mdH4Style,
                        MarkdownTextView_mdH4PostStyle),
                StyleGroup.of(a,
                        MarkdownTextView_mdH5PreStyle,
                        MarkdownTextView_mdH5Style,
                        MarkdownTextView_mdH5PostStyle),
                StyleGroup.of(a,
                        MarkdownTextView_mdH6PreStyle,
                        MarkdownTextView_mdH6Style,
                        MarkdownTextView_mdH6PostStyle)
        ));
        emphasisStyleGroup = StyleGroup.of(a,
                ID_NULL,
                MarkdownTextView_mdEmphasisStyle,
                ID_NULL);
        strongEmphasisStyleGroup = StyleGroup.of(a,
                ID_NULL,
                MarkdownTextView_mdStrongEmphasisStyle,
                ID_NULL);
        blockQuoteStyleGroup = StyleGroup.of(a,
                ID_NULL,
                MarkdownTextView_mdBlockQuoteStyle,
                ID_NULL);
        codeStyleGroup = StyleGroup.of(a,
                ID_NULL,
                MarkdownTextView_mdCodeStyle,
                ID_NULL);
        fencedCodeBlockStyleGroup = StyleGroup.of(a,
                ID_NULL,
                MarkdownTextView_mdFencedCodeBlockStyle,
                ID_NULL);
        indentedCodeBlockStyleGroup = StyleGroup.of(a,
                ID_NULL,
                MarkdownTextView_mdIndentedCodeBlockStyle,
                ID_NULL);
        orderedListItemStyleGroup = StyleGroup.of(a,
                ID_NULL,
                MarkdownTextView_mdOrderedListItemStyle,
                ID_NULL);
        unorderedListItemStyleGroup = StyleGroup.of(a,
                ID_NULL,
                MarkdownTextView_mdUnorderedListItemStyle,
                ID_NULL);
        paragraphStyleGroup = StyleGroup.of(a,
                ID_NULL,
                MarkdownTextView_mdParagraphStyle,
                ID_NULL);

        codeBlockBackgroundColor = a.getColorStateList(MarkdownTextView_mdCodeBlockBackgroundColor);
        codeBlockMargin = a.getDimensionPixelSize(MarkdownTextView_mdCodeBlockMargin, UNDEFINED_DIMEN);
        codeBlockTextColor = a.getColorStateList(MarkdownTextView_mdCodeBlockTextColor);

        if (a.hasValue(MarkdownTextView_mdLinkUnderlined)) {
            linkUnderlined = a.getBoolean(MarkdownTextView_mdLinkUnderlined, true);
            linkStyleGroup = StyleGroup.of(a,
                    ID_NULL,
                    MarkdownTextView_mdLinkStyle,
                    ID_NULL);
        } else {
            linkUnderlined = null;
            linkStyleGroup = StyleGroup.of(a,
                    ID_NULL,
                    ID_NULL,
                    MarkdownTextView_mdLinkStyle);
        }

        listItemColor = a.getColorStateList(MarkdownTextView_mdListItemColor);
        thematicBreakColor = a.getColorStateList(MarkdownTextView_mdThematicBreakColor);
        thematicBreakHeight = a.getDimensionPixelSize(MarkdownTextView_mdThematicBreakHeight, UNDEFINED_DIMEN);

        a.recycle();
    }

    static class StyleGroup {
        private static StyleGroup EMPTY = new StyleGroup(ID_NULL, ID_NULL, ID_NULL);

        @NonNull
        static StyleGroup of(@NonNull TypedArray attributes,
                             @StyleableRes int preStyleable,
                             @StyleableRes int styleable,
                             @StyleableRes int postStyleable) {
            final int preStyle = preStyleable == ID_NULL ? ID_NULL : attributes.getResourceId(preStyleable, ID_NULL);
            final int style = styleable == ID_NULL ? ID_NULL : attributes.getResourceId(styleable, ID_NULL);
            final int postStyle = postStyleable == ID_NULL ? ID_NULL : attributes.getResourceId(postStyleable, ID_NULL);
            if (preStyle == ID_NULL && style == ID_NULL && postStyle == ID_NULL) {
                return EMPTY;
            } else {
                return new StyleGroup(preStyle, style, postStyle);
            }
        }

        @StyleRes
        final int preStyle;

        @StyleRes
        final int style;

        @StyleRes
        final int postStyle;

        private StyleGroup(@StyleRes int preStyle,
                           @StyleRes int style,
                           @StyleRes int postStyle) {
            this.preStyle = preStyle;
            this.style = style;
            this.postStyle = postStyle;
        }

        final boolean hasPreStyle() {
            return isStyleValid(preStyle);
        }

        final boolean hasStyle() {
            return isStyleValid(style);
        }

        final boolean hasPostStyle() {
            return isStyleValid(postStyle);
        }

        static boolean isStyleValid(@StyleRes int styleResource) {
            return styleResource != ID_NULL;
        }
    }
}

File: library/src/main/java/se/ingenuity/markdownview/util/ResolvedAttributes.kt
package se.ingenuity.markdownview.util

import android.content.Context
import android.content.res.ColorStateList
import android.content.res.TypedArray
import android.util.AttributeSet
import androidx.annotation.AttrRes
import androidx.annotation.Dimension
import androidx.annotation.StyleRes
import androidx.annotation.StyleableRes
import se.ingenuity.markdownview.R

internal class ResolvedAttributes(
    context: Context,
    attrs: AttributeSet?,
    @AttrRes defStyleAttr: Int,
    @StyleRes defStyleRes: Int
) {
    val headingStyleGroups: List<StyleGroup>
    val emphasisStyleGroup: StyleGroup
    val strongEmphasisStyleGroup: StyleGroup
    val blockQuoteStyleGroup: StyleGroup
    val codeStyleGroup: StyleGroup
    val fencedCodeBlockStyleGroup: StyleGroup
    val indentedCodeBlockStyleGroup: StyleGroup
    val orderedListItemStyleGroup: StyleGroup
    val unorderedListItemStyleGroup: StyleGroup
    val paragraphStyleGroup: StyleGroup
    val linkStyleGroup: StyleGroup
    val linkUnderlined: Boolean?
    val listItemColor: ColorStateList?
    val codeBlockBackgroundColor: ColorStateList?
    val codeBlockTextColor: ColorStateList?

    @Dimension
    val codeBlockMargin: Int
    val thematicBreakColor: ColorStateList?

    @Dimension
    val thematicBreakHeight: Int

    init {
        val a = context.obtainStyledAttributes(
            attrs,
            R.styleable.MarkdownTextView,
            defStyleAttr,
            defStyleRes
        )
        headingStyleGroups = listOf(
            StyleGroup.of(
                a,
                R.styleable.MarkdownTextView_mdH1PreStyle,
                R.styleable.MarkdownTextView_mdH1Style,
                R.styleable.MarkdownTextView_mdH1PostStyle
            ),
            StyleGroup.of(
                a,
                R.styleable.MarkdownTextView_mdH2PreStyle,
                R.styleable.MarkdownTextView_mdH2Style,
                R.styleable.MarkdownTextView_mdH2PostStyle
            ),
            StyleGroup.of(
                a,
                R.styleable.MarkdownTextView_mdH3PreStyle,
                R.styleable.MarkdownTextView_mdH3Style,
                R.styleable.MarkdownTextView_mdH3PostStyle
            ),
            StyleGroup.of(
                a,
                R.styleable.MarkdownTextView_mdH4PreStyle,
                R.styleable.MarkdownTextView_mdH4Style,
                R.styleable.MarkdownTextView_mdH4PostStyle
            ),
            StyleGroup.of(
                a,
                R.styleable.MarkdownTextView_mdH5PreStyle,
                R.styleable.MarkdownTextView_mdH5Style,
                R.styleable.MarkdownTextView_mdH5PostStyle
            ),
            StyleGroup.of(
                a,
                R.styleable.MarkdownTextView_mdH6PreStyle,
                R.styleable.MarkdownTextView_mdH6Style,
                R.styleable.MarkdownTextView_mdH6PostStyle
            )
        )
        emphasisStyleGroup = StyleGroup.of(
            a,
            R.styleable.MarkdownTextView_mdEmphasisPreStyle,
            R.styleable.MarkdownTextView_mdEmphasisStyle,
            R.styleable.MarkdownTextView_mdEmphasisPostStyle
        )
        strongEmphasisStyleGroup = StyleGroup.of(
            a,
            R.styleable.MarkdownTextView_mdStrongEmphasisPreStyle,
            R.styleable.MarkdownTextView_mdStrongEmphasisStyle,
            R.styleable.MarkdownTextView_mdStrongEmphasisPostStyle
        )
        blockQuoteStyleGroup = StyleGroup.of(
            a,
            R.styleable.MarkdownTextView_mdBlockQuotePreStyle,
            R.styleable.MarkdownTextView_mdBlockQuoteStyle,
            R.styleable.MarkdownTextView_mdBlockQuotePostStyle
        )
        codeStyleGroup = StyleGroup.of(
            a,
            R.styleable.MarkdownTextView_mdCodePreStyle,
            R.styleable.MarkdownTextView_mdCodeStyle,
            R.styleable.MarkdownTextView_mdCodePostStyle
        )
        fencedCodeBlockStyleGroup = StyleGroup.of(
            a,
            R.styleable.MarkdownTextView_mdFencedCodeBlockPreStyle,
            R.styleable.MarkdownTextView_mdFencedCodeBlockStyle,
            R.styleable.MarkdownTextView_mdFencedCodeBlockPostStyle
        )
        indentedCodeBlockStyleGroup = StyleGroup.of(
            a,
            R.styleable.MarkdownTextView_mdIndentedCodeBlockPreStyle,
            R.styleable.MarkdownTextView_mdIndentedCodeBlockStyle,
            R.styleable.MarkdownTextView_mdIndentedCodeBlockPostStyle
        )
        orderedListItemStyleGroup = StyleGroup.of(
            a,
            R.styleable.MarkdownTextView_mdOrderedListItemPreStyle,
            R.styleable.MarkdownTextView_mdOrderedListItemStyle,
            R.styleable.MarkdownTextView_mdOrderedListItemPostStyle
        )
        unorderedListItemStyleGroup = StyleGroup.of(
            a,
            R.styleable.MarkdownTextView_mdUnorderedListItemPreStyle,
            R.styleable.MarkdownTextView_mdUnorderedListItemStyle,
            R.styleable.MarkdownTextView_mdUnorderedListItemPostStyle
        )
        paragraphStyleGroup = StyleGroup.of(
            a,
            R.styleable.MarkdownTextView_mdParagraphPreStyle,
            R.styleable.MarkdownTextView_mdParagraphStyle,
            R.styleable.MarkdownTextView_mdParagraphPostStyle
        )
        codeBlockBackgroundColor =
            a.getColorStateList(R.styleable.MarkdownTextView_mdCodeBlockBackgroundColor)
        codeBlockMargin = a.getDimensionPixelSize(
            R.styleable.MarkdownTextView_mdCodeBlockMargin,
            Constants.UNDEFINED_DIMEN
        )
        codeBlockTextColor = a.getColorStateList(R.styleable.MarkdownTextView_mdCodeBlockTextColor)
        linkUnderlined = if (a.hasValue(R.styleable.MarkdownTextView_mdLinkUnderlined)) {
            a.getBoolean(R.styleable.MarkdownTextView_mdLinkUnderlined, true)
        } else {
            null
        }
        linkStyleGroup = StyleGroup.of(
            a,
            R.styleable.MarkdownTextView_mdLinkPreStyle,
            R.styleable.MarkdownTextView_mdLinkStyle,
            R.styleable.MarkdownTextView_mdLinkPostStyle
        )
        listItemColor = a.getColorStateList(R.styleable.MarkdownTextView_mdListItemColor)
        thematicBreakColor = a.getColorStateList(R.styleable.MarkdownTextView_mdThematicBreakColor)
        thematicBreakHeight = a.getDimensionPixelSize(
            R.styleable.MarkdownTextView_mdThematicBreakHeight,
            Constants.UNDEFINED_DIMEN
        )
        a.recycle()
    }

    internal data class StyleGroup(
        @field:StyleRes @param:StyleRes val preStyle: Int,
        @field:StyleRes @param:StyleRes val style: Int,
        @field:StyleRes @param:StyleRes val postStyle: Int
    ) {
        fun hasPreStyle() = isStyleValid(preStyle)

        fun hasStyle() = isStyleValid(style)

        fun hasPostStyle() = isStyleValid(postStyle)

        companion object {
            private val EMPTY = StyleGroup(Constants.ID_NULL, Constants.ID_NULL, Constants.ID_NULL)

            fun of(
                attributes: TypedArray,
                @StyleableRes preStyleable: Int,
                @StyleableRes styleable: Int,
                @StyleableRes postStyleable: Int
            ): StyleGroup {
                val preStyle =
                    if (preStyleable == Constants.ID_NULL) Constants.ID_NULL else attributes.getResourceId(
                        preStyleable,
                        Constants.ID_NULL
                    )
                val style =
                    if (styleable == Constants.ID_NULL) Constants.ID_NULL else attributes.getResourceId(
                        styleable,
                        Constants.ID_NULL
                    )
                val postStyle =
                    if (postStyleable == Constants.ID_NULL) Constants.ID_NULL else attributes.getResourceId(
                        postStyleable,
                        Constants.ID_NULL
                    )

                return if (
                    preStyle == Constants.ID_NULL &&
                    style == Constants.ID_NULL &&
                    postStyle == Constants.ID_NULL
                ) {
                    EMPTY
                } else {
                    StyleGroup(preStyle, style, postStyle)
                }
            }

            fun isStyleValid(@StyleRes styleResource: Int): Boolean {
                return styleResource != Constants.ID_NULL
            }
        }
    }
}
